### PR TITLE
Tests: Improve configuration and mocking strategy

### PR DESCRIPTION
**Pull request by gziolo**
## Description
I encountered some issues with our current test setup when trying to debug failing tests in #8189.

This PR tries to improve our test configuration by:
* Removing global mocks which are enabled by default for every test suite. Instead, we enable all mocks explicitly using `jest.mock` calls.
* Fixing `moduleNameMapper` to always use source code rather than build files.
* Removing no longer needed `wp` global aliasing in the test setup.
* Removes `__mocks__` folder from the build output for all packages.

We should tackle #7005 next. It should be easier to identify tests which need to be updated since we can easily identify tests that depend on `Button` by searching for `jest.mock` call.

## How has this been tested?
* `npm run prebuild:packages` - to ensure there are no build files created for production code.
* `npm test` - to make sure all tests pass.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation.

### CODE REVIEW

**Comment by youknowriad** [comment]
Kind of unrelated, but can I ask why we need to mock the "Button" component, it's not really clear to me?

**Comment by gziolo** [comment]
> Kind of unrelated, but can I ask why we need to mock the "Button" component, it's not really clear to me?

See #7005 and #6527. In the latter, we wrapped `Button` with `forwardRef` which doesn't work with `enzyme` so we had to introduce such workaround. I hope to tackle it next week.

File: edit-post/components/header/more-menu/test/index.js [hunk 1]
```diff
@@ -8,6 +8,8 @@ import { mount } from 'enzyme';
  */
 import MoreMenu from '../index';
 
+jest.mock( '../../../../../packages/components/src/button' );
```
Comment (tofumatt) [inline edit-post/components/header/more-menu/test/index.js]: That is a nigh-unusable import path? I know there's discussion around just using absolute paths everywhere (I brought it up in a chat and should really file an issue... so: #8190).

But this should be absolute if possible. It's really unusable to try to figure out what this is importing.

File: packages/components/src/button/__mocks__/index.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+// [TEMPORARY]: Button uses React.forwardRef, added in react@16.3.0 but not yet
```
Comment (tofumatt) [inline packages/components/src/button/__mocks__/index.js]: Thanks for the docs! 👍

File: __mocks__/@wordpress/api-fetch/index.js [hunk 3]
```diff
@@ -0,0 +1,6 @@
+const apiFetch = jest.fn( () => {
+	return apiFetch.mockReturnValue;
+} );
+apiFetch.mockReturnValue = 'mock this value by overriding apiFetch.mockReturnValue';
```
Comment (tofumatt) [inline __mocks__/@wordpress/api-fetch/index.js]: Maybe make this a full sentence? `'Mock this value by overriding apiFetch.mockReturnValue.'`

**Review by tofumatt — APPROVED** [review]
So, I dig it and this is nicer, but those imports are quite something.

I filed #8190 to deal with them on a broader scale, but I'd really like it if the imports in here were absolute. Right now it's really tough to figure out where those files actually are located.

Code looks good though, just minor things to fix and then merge!

File: __mocks__/@wordpress/api-fetch/index.js (hunk 3 above)
Comment (gziolo) [reply __mocks__/@wordpress/api-fetch/index.js]: I copied it over - can tweak 👍

File: edit-post/components/header/more-menu/test/index.js (hunk 1 above)
Comment (gziolo) [reply edit-post/components/header/more-menu/test/index.js]: Agreed, it took me some time to build them properly 😅 

In the first place, I want to get rid of those mocks completely with #7005 which will sort of resolve this issue :D

File: package.json [hunk 4]
```diff
@@ -144,8 +144,8 @@
 	"scripts": {
 		"prebuild": "npm run check-engines",
 		"clean:packages": "rimraf ./packages/*/build ./packages/*/build-module ./packages/*/build-style",
-		"prebuild:packages": "npm run clean:packages && lerna run build && cross-env INCLUDE_PACKAGES=babel-plugin-import-jsx-pragma,postcss-themes SKIP_JSX_PRAGMA_TRANSFORM=1 node ./bin/packages/build.js",
-		"build:packages": "cross-env EXCLUDE_PACKAGES=babel-plugin-import-jsx-pragma,postcss-themes node ./bin/packages/build.js",
+		"prebuild:packages": "npm run clean:packages && lerna run build && cross-env INCLUDE_PACKAGES=babel-plugin-import-jsx-pragma,postcss-themes,jest-console SKIP_JSX_PRAGMA_TRANSFORM=1 node ./bin/packages/build.js",
```
Comment (aduth) [inline package.json]: Why are we including these other packages in the prebuild?

Comment (gziolo) [reply package.json]: * `postcss-themes` is necessary to build CSS styles for packages, so we need its source upfront
* `jest-console` doesn't strictly need to be there; however, I put it there to be able to run `npm run prebuild:package && npm test` and verify that packages can be tested without transpiled code

**Pull request merged**